android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt

```diff
@@ -0,0 +1,157 @@
+package com.badoo.ribs.android
+
+import android.annotation.TargetApi
+import android.content.pm.PackageManager
+import android.os.Build
+import android.support.v4.app.ActivityCompat
+import android.support.v4.content.ContextCompat
+import android.support.v7.app.AppCompatActivity
+import com.badoo.ribs.android.PermissionRequester.CheckPermissionsResult
+import com.badoo.ribs.android.PermissionRequester.RequestPermissionsEvent
+import com.badoo.ribs.android.PermissionRequester.RequestPermissionsEvent.Cancelled
+import com.badoo.ribs.android.PermissionRequester.RequestPermissionsEvent.RequestPermissionsResult
+import com.badoo.ribs.core.Identifiable
+import com.badoo.ribs.core.requestcode.RequestCodeRegistry
+import com.jakewharton.rxrelay2.PublishRelay
+import io.reactivex.Observable
+
+class PermissionRequesterImpl(
+    private val activity: AppCompatActivity,
+    private val requestCodeRegistry: RequestCodeRegistry
+) : PermissionRequester, PermissionRequestResultHandler {
+
+    private val events = HashMap<Int, HashMap<Int, PublishRelay<RequestPermissionsEvent>>>()
+
+    override fun checkPermissions(
+        client: Identifiable,
+        permissions: Array<String>
+    ) : CheckPermissionsResult {
+        val granted = mutableListOf<String>()
+        val canAsk = mutableListOf<String>()
+        val shouldShowRationale = mutableListOf<String>()
+
+        permissions.forEach { permission ->
+            if (ContextCompat.checkSelfPermission(activity, permission) == PackageManager.PERMISSION_GRANTED) {
+                granted.add(permission)
+            } else {
+                if (ActivityCompat.shouldShowRequestPermissionRationale(activity, permission)) {
+                    shouldShowRationale.add(permission)
+                } else {
+                    canAsk.add(permission)
+                }
+            }
+        }
+
+        return CheckPermissionsResult(
+            granted, canAsk, shouldShowRationale
+        )
+    }
+
+    @TargetApi(Build.VERSION_CODES.M)
+    override fun requestPermissions(
+        client: Identifiable,
+        requestCode: Int,
+        permissions: Array<String>
+    ) {
+        val code = requestCodeRegistry.generateRequestCode(client.id(), requestCode)
+
+        ActivityCompat.requestPermissions(activity,
+            permissions,
+            code
+        )
+    }
+
+    // TODO Consider what is better: with requestCode here (separate streams) or without (client code branching on single stream)
+    override fun events(client: Identifiable, requestCode: Int): Observable<RequestPermissionsEvent> {
+        val id = requestCodeRegistry.generateGroupId(client.id())
+        ensureSubject(id, requestCode)
+
+        return events[id]!![requestCode]!!
```

Comment:
Use getValue(id) instead of !! because it will throw more meaningful error (`no such key ...` instead of null pointer exception) and for me, it looks nicer than `!!`

Reply:
Done.

Reply:
Agreed with @lukaville.
In `PermissionRequesterImpl` file, we can replace accessing map value with `getValue`.